presto-main/src/main/java/com/facebook/presto/operator/aggregation/OptimizedTypedSet.java

```diff
@@ -0,0 +1,353 @@
+import com.facebook.presto.common.block.BlockBuilder;
+import com.facebook.presto.common.block.DictionaryBlock;
+import com.facebook.presto.common.block.DictionaryId;
+import com.facebook.presto.common.type.Type;
+import com.facebook.presto.operator.project.SelectedPositions;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import static com.facebook.presto.array.Arrays.ensureCapacity;
+import static com.facebook.presto.operator.project.SelectedPositions.positionsList;
+import static com.facebook.presto.type.TypeUtils.hashPosition;
+import static com.facebook.presto.type.TypeUtils.positionEqualsPosition;
+import static com.google.common.base.Preconditions.checkArgument;
+import static io.airlift.slice.SizeOf.sizeOf;
+import static it.unimi.dsi.fastutil.HashCommon.arraySize;
+import static java.lang.Math.max;
+import static java.lang.Math.toIntExact;
+import static java.util.Objects.requireNonNull;
+
+public class OptimizedTypedSet
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(TypedSet.class).instanceSize();
+    private static final int ARRAY_LIST_INSTANCE_SIZE = ClassLayout.parseClass(ArrayList.class).instanceSize();
+    private static final float FILL_RATIO = 0.75f;
+    private static final int EMPTY_SLOT = -1;
+    private static final int INVALID_POSITION = -1;
+    private static final int INITIAL_BLOCK_COUNT = 2;
+    private static final SelectedPositions EMPTY_SELECTED_POSITIONS = positionsList(new int[0], 0, 0);
+
+    private final String functionName;
+    private final Type elementType;
+    private final int hashCapacity;
+    private final int hashMask;
+
+    private int size;  // size is the number of elements added to the TypedSet (including null).
+    private Block[] blocks;   // Keeps track of the added blocks, even if the elements of the block was not inserted into the set. Array is used to get higher performance in getInsertPosition()
+    private List<SelectedPositions> positionsForBlocks;  // The selected positions for the added blocks, one for each added block
+    private long[] blockPositionByHash;  // Each 64-bit long is 32-bit index for blocks + 32-bit position within block
+    private int currentBlockIndex = -1;  // The index into the blocks array and positionsForBlocks list
+
+    public OptimizedTypedSet(Type elementType, int maxPositionCount, String functionName)
+    {
+        this(elementType, INITIAL_BLOCK_COUNT, maxPositionCount, functionName);
+    }
+
+    public OptimizedTypedSet(Type elementType, int expectedBlockCount, int maxPositionCount, String functionName)
+    {
+        checkArgument(expectedBlockCount >= 0, "expectedBlockCount must not be negative");
+        checkArgument(maxPositionCount >= 0, "maxPositionCount must not be negative");
+
+        this.elementType = requireNonNull(elementType, "elementType must not be null");
+        this.functionName = functionName;
```

Comment:
What is this for?

Reply:
@rongrong It used to be in `TypedSet` and used in the error message for `EXCEEDED_FUNCTION_MEMORY_LIMIT`. But the new implementation removed this limit so we don't need this `functionName` anymore. I have removed it. Thank you for the catch!